### PR TITLE
many: order structures in gadget according to offset

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Order structures in gadget according to offset as preparation for min-size and some minor refactoring.

### CODE REVIEW

File: gadget/gadget.go [hunk 1]
```diff
@@ -579,6 +582,62 @@ func compatWithPibootOrIndeterminate(m Model) bool {
...
+func orderStructuresByOffset(vss []VolumeStructure) []VolumeStructure {
+	if vss == nil {
+		return nil
+	}
+
+	// Build contiguous structures
+	scss := setContigStructs{}
+	var currentCont *contiguousStructs
+	for _, s := range vss {
+		// If offset we can start a new "block", otherwise the structure
```
Comment (pedronis) [inline gadget/gadget.go:615]: If offset is set?

File: gadget/gadget.go [hunk 2]
```diff
@@ -579,6 +582,62 @@ func compatWithPibootOrIndeterminate(m Model) bool {
 	return m == nil || m.Grade() != asserts.ModelGradeUnset
 }
 
+// Ancillary structs to sort structures
```
[the remark this answers is not in the document]
Comment (pedronis) [reply gadget/gadget.go:585]: probably this should mention min-size too

// Each volume structure belong to its own contiguousStructs unless its offset is made unknown by min-size use

or something like that

**Review by pedronis — COMMENTED** [review]
did a pass, it seems that are some changes/renamings that are not covered by the summary, I was a bit confused about them initially

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks for the review, comments addressed now. Sorry if I mixed up a couple things together here.

**Review by pedronis — APPROVED** [review]
thanks

File: gadget/gadget.go [hunk 3]
```diff
@@ -579,6 +582,63 @@ func compatWithPibootOrIndeterminate(m Model) bool {
 	return m == nil || m.Grade() != asserts.ModelGradeUnset
 }
 
+// Ancillary structs to sort volume structures. Each volume structure belong to
+// its own contiguousStructs unless its offset is made unknown by min-size use.
+type contiguousStructs struct {
```
Comment (Meulengracht) [inline gadget/gadget.go:587]: What does a contiguous structure represent in this context? Does it mean a list of ordered partitions by their offset on the volume, where a `contiguousStruct` represents a Volume?

File: gadget/update.go [hunk 4]
```diff
@@ -625,7 +625,7 @@ func DiskTraitsFromDeviceAndValidate(expLayout *LaidOutVolume, dev string, opts
 
 	// ensure that the on disk volume and the laid out volume are actually
 	// compatible
-	ensureOpts := &EnsureVolumeCompatibilityOptions{
+	ensureOpts := &VolumeCompatibilityOptions{
```
Comment (Meulengracht) [inline gadget/update.go:628]:
```suggestion
	volumeOpts := &VolumeCompatibilityOptions{
```

**Review by Meulengracht — COMMENTED** [review]
Question and a nitpick

File: gadget/gadget.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply gadget/gadget.go:592]: A contiguousStruct does not represent a Volume, a Volume would be composed of a set of contiguousStruct. Each contiguousStruct contains a first structure with a known fixed offset followed by structures for which the offset is unknown so we know for sure that they are after the first structure in contiguousStruct. I've expanded the comment to clarify this.

File: gadget/update.go (hunk 4 above)
Comment (alfonsosanchezbeato) [reply gadget/update.go]: Good suggestion, changed now

**Review by alfonsosanchezbeato — COMMENTED** [review]
@Meulengracht thanks for the review, I've addressed your comments now, and also rebased and cleaned-up history.

File: gadget/gadget.go (hunk 3 above)
Comment (Meulengracht) [reply gadget/gadget.go:592]: Thank you for the comment clarification, that helps

**Review by Meulengracht — APPROVED** [review]
Thank you for the changes, just a final suggestion on your updated comment

File: gadget/gadget.go [hunk 5]
```diff
@@ -579,6 +582,68 @@ func compatWithPibootOrIndeterminate(m Model) bool {
 	return m == nil || m.Grade() != asserts.ModelGradeUnset
 }
 
+// Ancillary structs to sort volume structures. We split volumes in
+// contiguousStructs slices, having each of these a first structure with a known
+// fixed offset followed by structures for which the offset is unknown so we
+// know for sure that they are after the first structure in
+// contiguousStruct. The unknown offsets appear becaus of min-size use. The
+// contiguousStructs are the things that we need to order, as all have a known
+// starting offset, which is not true for all the volume structures.
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply gadget/gadget.go]: Thanks, I've changed this & squashed in the previous commit

**Pull request merged**